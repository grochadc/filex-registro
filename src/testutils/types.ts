export interface Mock {
  request: { query: any; variables?: any };
  result: { errors?: any[]; data?: any };
}
