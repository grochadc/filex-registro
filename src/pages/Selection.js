import React, { useEffect, useState } from "react";
import { database, useFetch } from "../lib";
import { Formik } from "formik";
import * as Yup from "yup";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Jumbotron from "react-bootstrap/Jumbotron";

const FormSchema = Yup.object().shape({
  name: Yup.string()
    .min(2, "Too Short!")
    .max(50, "Too Long!")
    .required("Required"),
  firstLastName: Yup.string()
    .min(2, "Too Short!")
    .max(50, "Too Long!")
    .required("Required"),
  secondLastName: Yup.string()
    .min(2, "Too Short!")
    .max(50, "Too Long!")
    .required("Required"),
  code: Yup.number()
    .min(5, "Too short")
    .required("Required"),
  schedule: Yup.string().required("Porfavor elige tu horario")
});

const useAvailableSchedules = level => {
  const [schedules, setSchedules] = useState(null);
  useEffect(
    () => {
      if (level) {
        database
          .ref(`schedules/level${level}`)
          .once("value")
          .then(snapshot => {
            const availableSchedules = snapshot
              .val()
              .filter(schedule => schedule.registered < 35);
            setSchedules(availableSchedules);
          });
      }
    },
    [level]
  );
  return schedules;
};

function Selection(props) {
  const { response, loading, status } = useFetch(props.code, props.ubicacion);
  const student = response;
  let current_level;
  if (props.ubicacion) {
    current_level = student["NIVEL FINAL"];
  } else {
    current_level = student.pass ? student.prev_level + 1 : student.prev_level;
  }
  const schedule = useAvailableSchedules(current_level);
  delete student.id;
  return (
    <div>
      <Jumbotron>
        <Container>
          <h1> Bienvenido a Nivel {current_level}!</h1>
          <h2> Verifica que tus datos sean correctos y elige tu horario:</h2>
        </Container>
      </Jumbotron>
      <Container>
        {loading ? (
          <p>Cargando...</p>
        ) : status === 404 ? (
          <div>
            <p>Estudiante no encontrado</p>
            <Button
              variant="primary"
              onClick={() => props.handleRouteChange("home")}
            >
              Atrás
            </Button>
          </div>
        ) : status === 500 ? (
          <p>No se pudo conectar al servidor. Favor de intentar mas tarde.</p>
        ) : (
          <Formik
            initialValues={response}
            onSubmit={props.handleSubmit}
            validationSchema={FormSchema}
          >
            {({ values, errors, handleChange, handleSubmit }) => (
              <Form onSubmit={handleSubmit}>
                {Object.keys(values).map((key, index) => {
                  const labels = {
                    code: "Codigo",
                    name: "Nombre",
                    firstLastName: "Apellido Paterno",
                    secondLastName: "Appelido Materno",
                    ciclo: "Clico de Ingreso",
                    carrera: "Carrera",
                    genero: "Género",
                    email: "Correo Electronico",
                    celular: "Telefono Celular",
                    prev_level: "Nivel Anterior"
                  };
                  return key === "pass" ? null : props.ubicacion ? (
                    <Form.Group controlId={key} key={index}>
                      <Form.Label>{`${key}: `}</Form.Label>
                      <Form.Control
                        value={values[key]}
                        onChange={handleChange}
                        disabled={
                          (key === "ESCRITO") |
                          (key === "ORAL") |
                          (key === "NIVEL FINAL")
                        }
                      />
                    </Form.Group>
                  ) : (
                    <Form.Group controlId={key} key={index}>
                      <Form.Label>{`${labels[key]}: `}</Form.Label>
                      <Form.Control
                        value={values[key]}
                        onChange={handleChange}
                      />
                    </Form.Group>
                  );
                })}
                <Form.Group controlId="schedule">
                  <Form.Label>Elige tu horario:</Form.Label>
                  <Form.Control
                    as="select"
                    value={values.schedule}
                    onChange={handleChange}
                  >
                    <option value="">Horarios:</option>
                    {schedule ? (
                      schedule.map((item, index) => {
                        return (
                          <option key={index} value={item.group}>
                            {`${item.group} ${item.teacher} ${item.time}`}
                          </option>
                        );
                      })
                    ) : (
                      <option>Cargando horarios...</option>
                    )}
                  </Form.Control>
                  {errors && errors.schedule ? errors.schedule : null}
                </Form.Group>
                <Button variant="primary" type="submit" className="mb-3">
                  Enviar
                </Button>
              </Form>
            )}
          </Formik>
        )}
      </Container>
    </div>
  );
}

export default Selection;
