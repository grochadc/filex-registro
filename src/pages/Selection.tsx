import Alert from "react-bootstrap/Alert";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import { useFormik } from "formik";
import * as Yup from "yup";
import { useParams, useHistory } from "react-router-dom";
import { useQuery, useMutation, gql } from "@apollo/client";

const labels = {
  codigo: "Codigo:",
  nombre: "Nombre:",
  apellido_paterno: "Apellido Paterno:",
  apellido_materno: "Apellido Materno:",
  genero: "Genero:",
  carrera: "Carrera:",
  ciclo: "Ciclo de Ingreso:",
  telefono: "Telefono:",
  email: "Correo Electronico:",
  nivel: "Nivel:",
  curso: "Curso:",
};

export const GET_APPLICANT = gql`
  query info($codigo: ID!) {
    registeringLevel
    applicant(codigo: $codigo) {
      codigo
      nombre
      apellido_materno
      apellido_paterno
      genero
      carrera
      ciclo
      telefono
      email
      nivel
      curso
      externo
      nuevo_ingreso
      schedules {
        group
        serialized
      }
    }
  }
`;

export const REGISTER_STUDENT = gql`
  mutation register(
    $codigo: ID!
    $nombre: String!
    $apellido_materno: String!
    $apellido_paterno: String!
    $genero: String!
    $carrera: String!
    $ciclo: String!
    $telefono: String!
    $email: String!
    $nivel: String!
    $curso: String!
    $externo: Boolean!
    $nuevo_ingreso: Boolean!
    $schedule: String!
  ) {
    registerStudent(
      input: {
        codigo: $codigo
        nombre: $nombre
        apellido_materno: $apellido_materno
        apellido_paterno: $apellido_paterno
        genero: $genero
        carrera: $carrera
        ciclo: $ciclo
        telefono: $telefono
        email: $email
        nivel: $nivel
        curso: $curso
        externo: $externo
        nuevo_ingreso: $nuevo_ingreso
        grupo: $schedule
      }
    ) {
      nombre
      schedule {
        group
        teacher
        time
      }
    }
  }
`;

const composeInitialValues = (applicant) => {
  if (applicant === undefined) return [[], {}];
  const { schedules, __typename, ...rest } = applicant;
  return [schedules, rest];
};

const Selection = (props: { setMutationResponse: any }) => {
  const params: { code: string } = useParams();
  const history = useHistory();
  const query = useQuery(GET_APPLICANT, { variables: { codigo: params.code } });
  const { data } = query;
  const [schedules, initialValues] = composeInitialValues(
    data && data.applicant
  );
  const [registerStudent] = useMutation(REGISTER_STUDENT);
  const handleSubmit = (values) => {
    registerStudent({ variables: values })
      .then((res) => {
        props.setMutationResponse(res.data.registerStudent);
        history.push("/success");
      })
      .catch(console.error);
  };

  const Schema = Yup.object().shape({
    schedule: Yup.string().min(4).required("Elige un horario para continuar"),
  });
  const formik = useFormik({
    enableReinitialize: true,
    initialValues: { ...initialValues, schedule: "" },
    validationSchema: Schema,
    onSubmit: handleSubmit,
  });
  const isInvalidCode = /^[0-9]+$/.test(params.code) ? false : true;

  if (query.loading) return <p>Loading...</p>;
  if (query.error) return <>{JSON.stringify(query.error)}</>;
  if (isInvalidCode)
    return <Alert variant="danger">Ese no es un codigo valido.</Alert>;
  if (!data.registeringLevel.includes(Number(data.applicant.nivel)))
    return (
      <Alert variant="warning">
        Hoy no es dia de registro para nivel {data.applicant.nivel}. Por favor
        vuelve el dia indicado.
      </Alert>
    );
  if (schedules.length === 0)
    return (
      <Alert variant="warning">
        Lo sentimos, todos los grupos pra nivel {data.applicant.nivel} estan
        llenos.
      </Alert>
    );
  return (
    <div>
      <h1>Selection</h1>
      <p>{params.code}</p>
      <Form onSubmit={formik.handleSubmit}>
        {Object.keys(formik.values).map((el: string, index: any) => {
          if (["schedule", "nuevo_ingreso", "externo"].includes(el))
            return null;
          return (
            <Form.Group key={index}>
              <Form.Label>{labels[el]}</Form.Label>
              <Form.Control
                type="text"
                value={formik.values[el]}
                onChange={formik.handleChange}
                disabled={["codigo", "nivel"].includes(el)}
              />
            </Form.Group>
          );
        })}
        <Form.Group>
          <Form.Label>Horario Filex:</Form.Label>
          <Form.Control
            as="select"
            id="schedule"
            value={formik.values.schedule}
            onChange={formik.handleChange}
          >
            <option value="">Elige un horario:</option>
            {schedules.map((item, index: number) => (
              <option key={index} value={item.group}>
                {item.serialized}
              </option>
            ))}
          </Form.Control>
          {formik.errors.schedule && (
            <Alert variant="warning">{formik.errors.schedule}</Alert>
          )}
        </Form.Group>
        <Button type="submit">Enviar</Button>
      </Form>
    </div>
  );
};

export default Selection;
