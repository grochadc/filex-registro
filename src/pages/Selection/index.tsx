import React, { useEffect, useState } from "react";
import Alert from "react-bootstrap/Alert";
import { Loading, Error } from "../../components/utils";
import { useParams, useHistory } from "react-router-dom";
import { gql } from "@apollo/client";
import {
  UnenrolledStudent,
  useGetApplicantQuery,
  GetApplicantQuery,
} from "../../generated/grapqhl-types";
import { calculateCicloActual } from "../../utils";

import ApplicantEditor from "../../components/ApplicantEditor";
import ScheduleSelection from "../../components/ScheduleSelection";

export const GetApplicant = gql`
query GetApplicant($codigo: ID!, $cicloActual:String!) {
  unenrolledStudent(codigo: $codigo, cicloActual: $cicloActual) {
    codigo
    nombre
    apellido_materno
    apellido_paterno
    genero
    carrera
    cicloIngreso
    telefono
    email
    institucionalEmail
    nivel
    curso
    externo
    registering
    desertor
    registeredGroup {
      ciclo
      name
      time
      aula
      teacher
    }
    groups {
      id
      ciclo
      name
      time
      aula
      teacher
    }
  }
}
`;

type SelectionProps = {
  submitSelection: (studentWithSchedule: any) => void;
  onAlreadyRegistered: (info: { nombre: string; group: any }) => void;
};
const Selection = (props: SelectionProps) => {
  const [student, setStudent] = useState<
    GetApplicantQuery["unenrolledStudent"]
  >();

  const [groups, setGroups] = useState<
    GetApplicantQuery["unenrolledStudent"]["groups"]
  >([]);

  const [groupSelection, setGroupSelection] = useState("");

  //@ts-ignore
  const params: { code: string } = useParams();
  //const history = useHistory();
  const query = useGetApplicantQuery({
    variables: { codigo: params.code, cicloActual: calculateCicloActual(new Date()) },
    onCompleted: (data) => {
      if (data?.unenrolledStudent !== undefined) {
        setStudent(data.unenrolledStudent);
        setGroups(data.unenrolledStudent.groups);
      }
    },
  });
  const { data } = query;

  const [showScheduleSelection, setShowScheduleSelection] = useState(false);
  useEffect(() => { }, [showScheduleSelection]);

  const handleApplicantEditorSubmit = (values: any) => {
    setStudent({ ...student, ...values });
    setShowScheduleSelection(true);
  };

  const handleScheduleSelect = (index: number) => {
    const selectedGroup = groups[index];

    if (selectedGroup !== null && selectedGroup !== undefined) {
      props.submitSelection({
        codigo: student?.codigo,
        nivel: student?.nivel,
        curso: student?.curso,
        group: selectedGroup.name,
        groupId: selectedGroup.id,
      });
    }
  };
  const isInvalidCode = /^[0-9]+$/.test(params.code) ? false : true;

  useEffect(() => {
    if (data?.unenrolledStudent.registeredGroup) {
      console.log(
        "found alreadyRegistered",
        data.unenrolledStudent.registeredGroup
      );
      props.onAlreadyRegistered({
        nombre: data.unenrolledStudent.nombre,
        group: data.unenrolledStudent.registeredGroup,
      });
    }
  }, [data, props]);

  if (query.loading) return <Loading />;
  if(query.error) console.log(query.error);
  if (query.error) {
    return <Error err={query.error} />;
  }
  if (isInvalidCode)
    return <Alert variant="danger">Ese no es un codigo valido.</Alert>;
  if (data && !data.unenrolledStudent.registering)
    return (
      <Alert variant="warning">
        Hoy no es dia de registro para nivel {data.unenrolledStudent.nivel}. Por
        favor vuelve el dia indicado.
      </Alert>
    );
  if (data && data.unenrolledStudent.groups.length === 0)
    return (
      <Alert variant="warning">
        Lo sentimos, todos los grupos para nivel {data?.unenrolledStudent.nivel}{" "}
        estan llenos.
      </Alert>
    );
  if (student)
    return (
      <div>
        {student.desertor ? (
          <Alert variant="warning">
            Notamos que dertaste el ciclo anterior. Recuerda que solo puedes
            desertar dos veces!
          </Alert>
        ) : null}
        {showScheduleSelection ? (
          <ScheduleSelection
            schedules={groups}
            onScheduleSelect={handleScheduleSelect}
          />
        ) : (
          <ApplicantEditor
            type="edit"
            admin={false}
            heading="Revisa tus datos:"
            initialValues={student}
            onSubmit={handleApplicantEditorSubmit}
          />
        )}
      </div>
    );
  return <>No DATA</>;
};

export default Selection;
