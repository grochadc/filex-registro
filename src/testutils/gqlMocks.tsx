import { GET_APPLICANT, REGISTER_STUDENT } from "../queries";
export interface Mock {
  request: { query: any; variables?: any };
  result: { data: any };
}

const applicantInfo = {
  codigo: "1234567890",
  nombre: "Juan",
  apellido_materno: "Preciado",
  apellido_paterno: "Paramo",
  genero: "M",
  carrera: "Agronegocios",
  ciclo: "1955B",
  telefono: "3411234567",
  email: "juan@lamedialuna.net",
  nivel: "4",
  curso: "en",
  externo: false,
  schedule: "E4-1",
};

export const getApplicantMock: Mock = {
  request: { query: GET_APPLICANT, variables: { codigo: "1234567890" } },
  result: {
    data: {
      applicant: {
        ...applicantInfo,
        registering: true,
        schedules: [
          {
            teacher: "Gonzalo Rocha",
            group: "E4-1",
            serialized: "E4-1 Gonzalo Rocha",
          },
        ],
      },
    },
  },
};

export const registerStudentMock: Mock = {
  request: {
    query: REGISTER_STUDENT,
    variables: { ...applicantInfo, schedule: "E4-1" },
  },
  result: {
    data: {
      registerStudent: {
        nombre: applicantInfo.nombre,
        schedule: {
          group: "E4-1",
          teacher: "Gonzalo Rocha",
        },
      },
    },
  },
};
