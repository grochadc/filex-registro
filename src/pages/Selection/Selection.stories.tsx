import React from "react";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import generateMock from "../../testutils/generatedMocks";
import { MockedProvider } from "@apollo/client/testing";
import { MemoryRouter, Route } from "react-router-dom";

import Selection, { RegisterStudent, GetApplicant } from "./index";
export default {
  title: "Pages/Selection",
  component: Selection,
  argTypes: {
    setMutationResponse: { action: "submitted" },
    onAlreadyRegistered: { action: "registered" },
    submitSelection: { action: "selection" },
  },
  parameters: {
    apolloClient: {
      mocks: [generateMock(RegisterStudent, {}, {})],
      MockedProvider,
    },
  },
  decorators: [
    (Story) => (
      <MemoryRouter initialEntries={["/selection/1234567890"]}>
        <Route path="/selection/:code">
          <Story />
        </Route>
      </MemoryRouter>
    ),
  ],
} as ComponentMeta<typeof Selection>;

const Template: ComponentStory<typeof Selection> = (args) => (
  <Selection {...args} />
);

const baseApplicantForMocks = {
  codigo: "1234567890",
  email: "email@website.com",
  ciclo: "2022A",
  telefono: "3411234567",
  curso: "en",
  genero: "M",
  nivel: "4",
  externo: true,
  registering: true,
  schedules: [
    { teacher: "Gonzalo", group: "E4-1" },
    { teacher: "Alondra", group: "E4-2" },
  ],
};

export const Normal = Template.bind({});
Normal.parameters = {
  apolloClient: {
    mocks: [
      generateMock(
        GetApplicant,
        { codigo: "1234567890" },
        {
          data: {
            applicant: baseApplicantForMocks,
          },
        }
      ),
      generateMock(
        RegisterStudent,
        {
          codigo: "1234567890",
          nombre: "Applicant-nombre",
          apellido_materno: "Applicant-apellido_materno",
          apellido_paterno: "Applicant-apellido_paterno",
          genero: "M",
          carrera: "Applicant-carrera",
          ciclo: "2022A",
          telefono: "3411234567",
          email: "email@website.com",
          nivel: "4",
          curso: "en",
          externo: true,
          schedule: "E4-1",
        },
        {}
      ),
    ],
  },
};

export const WrongDay = Template.bind({});
WrongDay.parameters = {
  apolloClient: {
    mocks: [
      generateMock(
        GetApplicant,
        { codigo: "1234567890" },
        {
          data: {
            applicant: {
              ...baseApplicantForMocks,
              registering: false,
            },
          },
        }
      ),
    ],
  },
};

export const LevelsFull = Template.bind({});
LevelsFull.parameters = {
  apolloClient: {
    mocks: [
      generateMock(
        GetApplicant,
        { codigo: "1234567890" },
        {
          data: {
            applicant: { ...baseApplicantForMocks, schedules: [] },
          },
        }
      ),
    ],
  },
};
