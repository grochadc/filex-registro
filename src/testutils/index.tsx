import React from "react";
import { render } from "@testing-library/react";
import { getApplicantMock, registerStudentMock } from "./gqlMocks";
import { MockedProvider } from "@apollo/client/testing";

const AllTheProviders = ({ children }) => {
  return (
    <MockedProvider
      mocks={[getApplicantMock, registerStudentMock]}
      addTypename={false}
    >
      {children}
    </MockedProvider>
  );
};

const renderWithGql = (ui) => render(ui, { wrapper: AllTheProviders });

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderWithGql };
