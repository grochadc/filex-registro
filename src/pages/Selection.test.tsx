import { renderWithGql, screen, act } from "../testutils";
import Selection from "./Selection";
import { Router, Route } from "react-router-dom";
import { createMemoryHistory } from "history";

test("Selection", async () => {
  const history = createMemoryHistory();
  history.push("/selection/1234567890");
  const { asFragment } = renderWithGql(
    <Router history={history}>
      <Route path="/selection/:code">
        <Selection setMutationResponse={console.log} />
      </Route>
    </Router>
  );
  await act(async () => await new Promise((resolve) => setTimeout(resolve, 0)));

  expect(asFragment()).toMatchSnapshot();
});
