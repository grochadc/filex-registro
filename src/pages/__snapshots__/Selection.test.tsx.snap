// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Selection
    </h1>
    <p>
      1234567890
    </p>
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Codigo:
        </label>
        <input
          class="form-control"
          disabled=""
          type="text"
          value="1234567890"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Nombre:
        </label>
        <input
          class="form-control"
          type="text"
          value="Juan"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Apellido Materno:
        </label>
        <input
          class="form-control"
          type="text"
          value="Preciado"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Apellido Paterno:
        </label>
        <input
          class="form-control"
          type="text"
          value="Paramo"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Genero:
        </label>
        <input
          class="form-control"
          type="text"
          value="M"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Carrera:
        </label>
        <input
          class="form-control"
          type="text"
          value="Agronegocios"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Ciclo de Ingreso:
        </label>
        <input
          class="form-control"
          type="text"
          value="1955B"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Telefono:
        </label>
        <input
          class="form-control"
          type="text"
          value="3411234567"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Correo Electronico:
        </label>
        <input
          class="form-control"
          type="text"
          value="juan@lamedialuna.net"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Nivel:
        </label>
        <input
          class="form-control"
          disabled=""
          type="text"
          value="4"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Curso:
        </label>
        <input
          class="form-control"
          type="text"
          value="en"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          Horario Filex:
        </label>
        <select
          class="form-control"
          id="schedule"
        >
          <option
            value=""
          >
            Elige un horario:
          </option>
          <option
            value="E4-1"
          >
            E4-1 Gonzalo Rocha 07:00-08:00
          </option>
        </select>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Enviar
      </button>
    </form>
  </div>
</DocumentFragment>
`;
